w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+            except TypeError:
+                self.precedence == UNK
+
+            if reflection == rp:
+                self.precedence = CONCAT
+                desc = u'Concatenation of values of the two input parameters ' \
+                       u'with the same name \"%s\" detected inside HTML ' \
+                       u'code of response body.\nThe request url was %s.\n' \
+                       u'Got concatenated output reflected in this link %s\n' \
+                       u'This \"feature\" may be used in different WAF ' \
+                       u'bypassing techniques.'
+                desc = desc % (param, ptest_url, link)
+                print desc
+
+                i = Info(u'Input concatenation detected', desc, ptest_resp.id,
+                         self.get_name())
+                i.set_url(ptest_url)
+                self.kb_append(self, 'hpp', i)
+                return
+
+
+
+        if self.precedence == UNK:
+            om.out.debug(u'Unable to determine precedence for HPP detection.'
+                         u'Presumably not vulnerable to HPP.')
+
+            return
+#        print "Attack!!", self.precedence
+        om.out.debug(u"Detected precedence is %s."
+                     u"\nTrying pollution of the reflected parameters: %s."
+                     % (self.precedence, rparams.keys()))
+        #bulding payloads
+        payloads = []
+        vuln_params=[]
+        for k,v in rparams.items():
+            payloads.append(REFLECT_TOKEN+'&'+v+'='+FORGE_TOKEN)
+        vtest_mutants = create_mutants(freq, payloads,
+                                       fuzzable_param_list=rparams.keys(),
+                                       mutant_tuple=(QSMutant,))
+        for m in vtest_mutants:
+            vtest_resp = self._uri_opener.send_mutant(m)
+            try:
+                vtest_dp = parser_cache.dpc.get_document_parser_for(vtest_resp)
+            except BaseFrameworkException:
+                return
+            r =  self.find_reflection( vtest_dp, FORGE_TOKEN)
+            if r:
+                vuln_params.append(r)
+                msg=u"value of parameter %s can be forged with " \
+                    u"HPP attack using URL: %s, precedence is %s"
+                msg = msg % (r, m.get_uri(), self.precedence)
+                v=Vuln.from_mutant(u'HPP', desc=msg, mutant = m,
+                                   plugin_name='hpp',response_ids=[],
+                                   severity=severity.MEDIUM)
+                v.set_url(self._orig_uri)
+                v[HPPInfoSet.ITAG] = m.get_uri()
+                # debug_msg = debug_msg % (rid, freq.get_uri())
+                self.kb_append_uniq_group(self, 'hpp', v,
+                                          group_klass=HPPInfoSet)
+
+    def find_reflection(self, parser, pattern1, pattern2 = ''):
+        #finds reflection of a string or two concatenated
```

Comment:
Change `#` commend below find_reflection definition to a docstring, so it gets picked up by any tools that create source code documentation